From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: olek.white.green@gmail.com
Date: Mon, 10 Oct 2022 00:01:16 -0500
Subject: fix: Horizontal wheel event.


diff --git a/ui/views/controls/scrollbar/scroll_bar.cc b/ui/views/controls/scrollbar/scroll_bar.cc
index d5f7809939377..5c9f594fd3fa9 100644
--- a/ui/views/controls/scrollbar/scroll_bar.cc
+++ b/ui/views/controls/scrollbar/scroll_bar.cc
@@ -113,7 +113,14 @@ bool ScrollBar::OnKeyPressed(const ui::KeyEvent& event) {
 }
 
 bool ScrollBar::OnMouseWheel(const ui::MouseWheelEvent& event) {
-  OnScroll(event.x_offset(), event.y_offset());
+  // OnScroll(event.x_offset(), event.y_offset());
+  int x_offset = event.x_offset();
+#if BUILDFLAG(IS_WIN)
+  if (event.native_event().message == WM_MOUSEHWHEEL) {
+    x_offset = -x_offset;
+  }
+#endif
+  OnScroll(x_offset, event.y_offset());
   return true;
 }
 
